dask/dataframe/shuffle.py

```diff
@@ -183,6 +183,11 @@ def set_index(
     """See _Frame.set_index for docstring"""
     if isinstance(index, Series) and index._name == df.index._name:
         return df
+
+    if is_index_like(index):
+        index = index.to_series()
+        index.index = df.index
```

Comment:
This feels like cheating. Scrutinize this

Reply:
Hey @gjoseph92 -- thanks for the comments. To get to this (disappointingly short) diff, I messed around trying to figure out how to the `divisions` and `npartitions` could let us take a happier-path. But all of my incantations were very case-specific and didn't work for any index type (e.g., a RangeIndex). 

What's these two lines effectively do check that the new index is the correct length and then pretend it's already a column of the dataframe. Without `index.index = df.index`, nothing really worked. That might be b/c I was missing something fundamental.